rfcs/text/0011_global_search.md

```diff
@@ -0,0 +1,709 @@
+- Start Date: 2020-04-19
+- RFC PR: [#64284](https://github.com/elastic/kibana/pull/64284)
+- Kibana Issue: [#61657](https://github.com/elastic/kibana/issues/61657)
+
+# Summary
+
+A Kibana API exposed on both public and server side, allowing consumers to search for various objects and
+register result providers.
+
+# Basic example
+
+- registering a result provider:
+
+```ts
+coreSetup.globalSearch.registerResultProvider({
+  id: 'my_provider',
+  find: (term, options, context) => {
+    const resultPromise = myService.search(term, context.core.savedObjects.client);
+    return from(resultPromise);
+  },
+});
+```
+
+- using the `find` API from the client-side:
+
+```ts
+coreStart.globalSearch.find('some term').subscribe(({ results }) => {
+    updateResults(results);
+}, () => {}, () => {
+    showAsyncSearchIndicator(false);
+});
+```
+
+# Motivation
+
+Kibana should do its best to assist users searching for and navigating to the various objects present on the Kibana platform.
+
+We should expose an API to make it possible for plugins to search for the various objects present on a Kibana instance.
+
+The first consumer of this API will be the global search bar [#57576](https://github.com/elastic/kibana/issues/57576). This API
+should still be generic to answer similar needs from any other consumer, either client or server side.
+
+# Detailed design
+
+## API Design
+
+### Result provider API
+
+#### common types
+
+```ts
+/**
+ * Static, non exhaustive list of the common search types.
+ * Only present to allow consumers and result providers to have aliases to the most common types.
+ */
+enum GlobalSearchCommonResultTypes {
+  application = 'application,
+  dashboard = 'dashboard',
+  visualization = 'visualization',
+  search = 'search',
+}
+
+/**
+ * Options provided to {@link GlobalSearchResultProvider | result providers} `find` method.
+ */
+interface GlobalSearchProviderFindOptions {
+  /**
+   * A custom preference token associated with a search 'session' that should be used to get consistent scoring
+   * when performing calls to ES. Can also be used as a 'session' token for providers returning data from elsewhere
+   * than an elasticsearch cluster.
+   */
+  preference: string;
+  /**
+   * Observable that emit once if and when the `find` call has been aborted by the consumer, or when the timeout period as been reached.
+   * When a `find` request is aborted, the service will stop emitting any new result to the consumer anyway, but
+   * this can (and should) be used to cancel any pending asynchronous task and complete the result observable.
+   */
+  aborted$: Observable<void>;
```

Comment:
In `data` and `expressions` plugins we use [`AbortController` API](https://developer.mozilla.org/en-US/docs/Web/API/AbortController) for handling abortion signalling. It stores the state boolean whether request was aborted and one can attach event listeners to listen for change in that state.

Reply:
If you stick with observable, maybe it is worth making it `BehaviorSubject` or observable that emits value immediately on subscription if request was aborted, as user might immediately abort a request they started.

Reply:
Observables already have the concept of being "aborted" baked in. When a consumer unsubscribes it signals to the producer that it no longer cares about the result. This doesn't mean the producer **has** to abort any async work, but this is the default for many API's e.g. `ajax.getJSON`

If a consumer wants to make cancellation more explicit they can implement their own `aborted$` subject and use it to unsubscribe / cancel async requests:

```ts
import { ajax } from 'rxjs/ajax';
const aborted$ = new Subject();
ajax.getJSON(`/api/users/${action.payload}`).pipe(takeUntil(aborted$)).subscribe(/* do something */)
// Will cancel the XHR request if it's still pending
aborted$.next();
```

Reply:
So, does this mean there will be two ways to abort the search: (1) through unsubscription; (2) through `aborted$` observable.

Reply:
We are quite inconsistent regarding cancelation in core.

on the client-side `http.fetch` uses the `signal` parameter which is an `AbortSignal`
on the server-side, `KibanaRequest` has the `events.aborted$` property which is an `Observable` with a replay effect.

We can leverage observable unsubscription, but that would be yet another way to perform cancelation. Also it would imho be less 'explicit' for result providers that they should perform their cleanup at cancelation (aka `This doesn't mean the producer has to abort any async work, but this is the default for many API'`)

I still think i'd prefer using an explicit cancelation mechanism. Now, I don't have a very strong opinion between a signal or an observable.

> If you stick with observable, maybe it is worth making it BehaviorSubject or observable that emits value immediately on subscription if request was aborted

consumer should provide a cold observable and/or with a replay effect if they think there is a risk for the abort signal to be already emitted. This is afaik an implementation detail from the consumer's side. Using a Replay/BehaviorSubject is indeed an option for them, and the `Observable` type we are expecting would allow them to do so.